package/gluon-mesh-vpn-wireguard/files/lib/netifd/proto/gluon_wireguard.sh

```diff
@@ -0,0 +1,89 @@
+#!/bin/sh
+# Copyright 2016-2017 Christof Schulze <christof@christofschulze.com>
+# Licensed to the public under the Apache License 2.0.
+
+. /lib/functions.sh
+. ../netifd-proto.sh
+init_proto "$@"
+
+proto_gluon_wireguard_init_config() {
+  no_device=1
+  available=1
+  renew_handler=1
+}
+
+proto_gluon_wireguard_renew() {
+  local config="$1"
+  echo wireguard RENEW: $@
+  ifdown $config
+  ifup $config
+}
+
+proto_gluon_wireguard_setup() {
+  local config="$1"
+  ifname=$(uci get network.$config.ifname) # we need uci here because nodevice=1 means the device is not part of the ubus structure
+
+  if [[ $(wg show all latest-handshakes |wc -l) -ge $(uci get wireguard.mesh_vpn_backbone.peer_limit) ]]; then
+    echo "not establishing another connection, we already have  $(uci get wireguard.mesh_vpn_backbone.peer_limit) connections." >&2
+    ip link del $ifname
+    ifdown $config
+    exit 1
+  fi
+
+  (
+  flock -n 9
+
+  if [[ $(uci get wireguard.mesh_vpn_backbone.enabled) -eq 1 ]]; then
+    ip link del "$ifname"
+    ip link add dev "$ifname" type wireguard
+    ip link set mtu $(uci get network.$config.mtu) dev "$ifname"
+    ip link set multicast on dev "$ifname"
+
+    group_peers=$(uci show wireguard|grep mesh_vpn_backbone_peer|grep =peer|cut -d"." -f2|cut -d= -f1)
```

Comment:
- Don't use `uci show`
- Do you have an example how the wireguard config looks in UCI? I don't think we needs groups at all for now.

Reply:
the groups are needed to allow for the same flexibility as fastd. They are also needed for n2n vpn which there already is code for.

Reply:
I am not aware of any other mechanism to retrieve the backbone peer with shell. With lua it would be a different story.

This is the structure:
wireguard.mesh_vpn_backbone=peer_group
wireguard.mesh_vpn_backbone.enabled='1'
wireguard.mesh_vpn_backbone.peer_limit='2'
wireguard.mesh_vpn_backbone.net='mesh_vpn'
wireguard.mesh_vpn_backbone_peer_gw02=peer
wireguard.mesh_vpn_backbone_peer_gw02.enabled='1'
wireguard.mesh_vpn_backbone_peer_gw02.ifname='mesh-vpn-0'
wireguard.mesh_vpn_backbone_peer_gw02.key='BjwoigPuHD3Cyv7UEkPUqvQU06SKx8MAgIMRpN8vjFw='
wireguard.mesh_vpn_backbone_peer_gw02.group='mesh_vpn_backbone'
wireguard.mesh_vpn_backbone_peer_gw02.net='mesh_vpn'
wireguard.mesh_vpn_backbone_peer_gw02.broker_port='40000'
wireguard.mesh_vpn_backbone_peer_gw02.remote='gw02.babel.md.freifunk.net'
wireguard.mesh_vpn_backbone_peer_gw02.allowedIPs='::/0'
wireguard.mesh_vpn_backbone_peer_gw01=peer
wireguard.mesh_vpn_backbone_peer_gw01.enabled='1'
wireguard.mesh_vpn_backbone_peer_gw01.ifname='mesh-vpn-1'
wireguard.mesh_vpn_backbone_peer_gw01.key='2jixtbjIghjN5MLWG5dgqvyHVV6tZRqFV/EdVs3egnI='
wireguard.mesh_vpn_backbone_peer_gw01.group='mesh_vpn_backbone'
wireguard.mesh_vpn_backbone_peer_gw01.net='mesh_vpn'
wireguard.mesh_vpn_backbone_peer_gw01.broker_port='40000'
wireguard.mesh_vpn_backbone_peer_gw01.remote='gw01.babel.md.freifunk.net'
wireguard.mesh_vpn_backbone_peer_gw01.allowedIPs='::/0'
wireguard.mesh_vpn=mesh_vpn
wireguard.mesh_vpn.group='mesh_vpn_backbone'
wireguard.mesh_vpn.packet_mark='1'
wireguard.mesh_vpn.secret='geheim'